cache/cache.go

```diff
@@ -13,3 +13,284 @@
+		for {
+			select {
+			case event, ok := <-stream:
+				if !ok {
+					return
+				}
+				responseChan <- clientv3.WatchResponse{Events: []*clientv3.Event{event}}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}()
+	return responseChan
+}
+
+// Ready reports whether the cache has finished its initial load.
+func (c *Cache) Ready() bool {
+	select {
+	case <-c.ready:
+		return true
+	default:
+		return false
+	}
+}
+
+// WaitReady blocks until the cache is ready or the ctx is cancelled.
+func (c *Cache) WaitReady(ctx context.Context) error {
+	select {
+	case <-c.ready:
+		return nil
+	case <-ctx.Done():
+		return ctx.Err()
+	}
+}
+
+// Close cancels the private context and blocks until all goroutines return.
+func (c *Cache) Close() {
+	c.stop()
+	c.waitGroup.Wait()
+}
+
+// newWatchEventStream builds a single internal Watch event stream for
+// both historical replay (>= startRev) and live updates filtered by key.
+func (c *Cache) newWatchEventStream(ctx context.Context, key string, startRev int64) (WatchEventStream, error) {
+	var pred KeyPredicate
+	if key == "\x00" { // special case for "\x00" which means “entire key-space”
+		pred = nil // accept every event
+	} else {
+		pred = func(k []byte) bool { return strings.HasPrefix(string(k), key) }
+	}
+
+	// TODO: re-enable arbitrary startRev support once we guarantee gap-free replay.
+	if startRev != 0 {
+		return nil, ErrUnsupportedStartRev
+	}
+
+	var nextRev int64
+	if latestEvent := c.history.PeekLatest(); latestEvent != nil {
+		nextRev = latestEvent.Kv.ModRevision + 1
+	}
+
+	w := newWatcher(c.cfg.PerWatcherBufferSize, pred)
+	c.demux.add(w, nextRev)
+
+	go func() {
+		<-ctx.Done()
+		c.demux.remove(w)
+	}()
+
+	return w.eventQueue, nil
+}
+
+func serveWatchEvents(ctx context.Context, watchCtx *watchCtx, bufSize int) {
+	enqueueCh := make(chan *clientv3.Event, bufSize)
+
+	go func() {
+		for event := range enqueueCh {
+			watchCtx.broadcast(event)
+		}
+	}()
+
+	// upstream watch/retry loop
+	go func() {
+		defer close(enqueueCh)
+		watchRetryLoop(ctx, watchCtx, enqueueCh)
+	}()
+}
+
+func watchRetryLoop(ctx context.Context, watchCtx *watchCtx, enqueueCh chan<- *clientv3.Event) {
+	backoff := watchCtx.backoffStart
+	for {
+		opts := []clientv3.OpOption{
+			clientv3.WithPrefix(),
+			clientv3.WithProgressNotify(),
+			clientv3.WithCreatedNotify(),
+		}
+		if oldestEvent := watchCtx.ring.PeekOldest(); oldestEvent != nil {
+			opts = append(opts,
+				clientv3.WithRev(oldestEvent.Kv.ModRevision+1))
+		}
+		watchCh := watchCtx.client.Watch(ctx, watchCtx.prefix, opts...)
+
+		if drainWatchChannel(ctx, watchCh, watchCtx.ring, watchCtx.demux, enqueueCh, watchCtx.onFirstResponse) == nil {
+			return
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.After(backoff):
+		}
+		if backoff < watchCtx.backoffMax {
+			backoff *= 2
+		}
+	}
+}
+
+// drainWatchChannel reads an etcd Watch stream into History and enqueueCh, returning nil on cancel or first error.
+func drainWatchChannel(
```

Comment:
`drain` implies there is something finate to be emptied. Based on the comment maybe `read` would be better verb?